src/ripple/protocol/impl/LedgerFormats.cpp

```diff
@@ -165,4 +170,49 @@ LedgerFormats::getInstance ()
     return instance;
 }
 
+std::pair<RPC::Status, LedgerEntryType>
+chooseLedgerEntryType(Json::Value const& params)
+{
+    std::pair<RPC::Status, LedgerEntryType> result{RPC::Status::OK, ltINVALID};
+    if (params.isMember(jss::type))
+    {
+        static
+        std::array<std::pair<char const *, LedgerEntryType>, 9> const types
+            {{
+                { jss::account,     ltACCOUNT_ROOT  },
+                { jss::amendments,  ltAMENDMENTS    },
+                { jss::directory,   ltDIR_NODE      },
+                { jss::fee,         ltFEE_SETTINGS  },
+                { jss::hashes,      ltLEDGER_HASHES },
+                { jss::offer,       ltOFFER         },
+                { jss::signer_list, ltSIGNER_LIST   },
+                { jss::state,       ltRIPPLE_STATE  },
+                { jss::ticket,      ltTICKET        }
+            }};
+
+        auto const& p = params[jss::type];
+        if (!p.isString())
+        {
+            result.first = RPC::Status{rpcINVALID_PARAMS,
+                                       "Invalid field 'type', not string."};
+            assert(result.first.type() == RPC::Status::Type::error_code_i);
+            return result;
+        }
+
+        auto const filter = p.asString ();
+        auto iter = std::find_if (types.begin (), types.end (),
+            [&filter](decltype (types.front ())& t)
+                { return t.first == filter; });
+        if (iter == types.end ())
+        {
+            result.first = RPC::Status{rpcINVALID_PARAMS,
+                                       "Invalid field 'type'."};
```

Comment:
Yeah, I don't disagree with you.  However this is a refactoring of the generation of an existing error message, not a new one, and there is even an existing test looking for this exact message.  So for this PR I want to just add the new functionality without changing existing functionality.  If we want, we can improve error messages in a separate PR.

The rationale for the refactoring is so that this existing code (moved from src/ripple/rpc/handlers/AccountObjects.cpp) could do double-duty:  support its original use case and the new use case this PR introduces.